src/full/Agda/TypeChecking/Rules/LHS.hs

```diff
@@ -1717,12 +1695,23 @@ disambiguateConstructor ambC@(AmbQ cs) d pars = do
 
         return (c, con, cType)
 
-prettyDisamb :: QName -> TCM Doc
-prettyDisamb x = do
-  let d  = pretty =<< dropTopLevelModule x
-  let mr = lastMaybe $ filter (noRange /=) $ map nameBindingSite $ mnameToList $ qnameModule x
-  caseMaybe mr d $ \ r -> d <+> ("(introduced at " <> prettyTCM r <> ")")
+    -- This deduplication identifies different names of the same constructor, ensuring
+    -- that the "ambiguous constructor" error does not fire for the case described
+    -- in #4130.
+    dedupCons :: forall a b. [(a, ConHead, b)] -> [(a, ConHead, b)]
+    dedupCons = List.nubBy ((==) `on` conName . snd3)
```

Comment:
Unfortunately, this deduplication is too crude as there can be two instantiations of the same constructor that create an ambiguous constructor, see #4135.

Reply:
Not too sure how 4135 is related - there's no "ambiguous constructor" error in that example even if I remove this function.

In any case, the type of `dedupCons` can be made more concrete:
```
dedupCons :: [(QName, ConHead, Type)] -> [(QName, ConHead, Type)]
```
If the same constructor is instantiated in two different ways, I guess it will be reflected in the third component `Type`. If so we can dedup on both `ConHead` and `Type`. But I don't have a concrete example that I can use to verify this.

Reply:
I am refining this function in the course of #4135.

Reply:
Wow that's not a trivial fix at all. Apologies for the trouble!

Reply:
No need to apologize.  This wasn't a problem before I attempted #4135.